consistent-function-scoping false positive for uninitialized let variable

Running ESLint 6.8.0 with eslint-plugin-unicorn 17.1.0 and `"extends": "plugin:unicorn/recommended"` on a file with the following reduced example code:

```js
'use strict';

function cat(stream) {
  let content;

  function onEnd() {
    process.stdout.write(content);
  }

  stream
    .setEncoding('utf8')
    .on('data', (chunk) => {
      if (!content) {
        content = chunk;
      } else {
        content += chunk;
      }
    })
    .once('end', onEnd);
}

cat(process.stdin);
```
gives the following output:

    /path/to/example.js
      6:3  error  Move function to the outer scope  unicorn/consistent-function-scoping
    
    ✖ 1 problem (1 error, 0 warnings)

The error seems incorrect since the `onEnd` function can't be moved to the outer scope because it references `content` defined in the inner scope.

Note that if `let content;` is changed to `let content = '';` the error goes away.

Thanks,
Kevin

--- a/rules/consistent-function-scoping.js
+++ b/rules/consistent-function-scoping.js
@@ -12,7 +12,22 @@ const isSameScope = (scope1, scope2) =>
 	scope1 && scope2 && (scope1 === scope2 || scope1.block === scope2.block);
 
 function checkReferences(scope, parent, scopeManager) {
-	const hitReference = references => references.some(reference => isSameScope(parent, reference.from));
+	const hitReference = references => references.some(reference => {
+		if (isSameScope(parent, reference.from)) {
+			return true;
+		}
+
+		const {resolved} = reference;
+		const [definition] = resolved.defs;
+
+		// Skip recursive function name
+		if (definition && definition.type === 'FunctionName' && resolved.name === definition.name.name) {
+			return false;
+		}
+
+		return isSameScope(parent, resolved.scope);
+	});
+
 	const hitDefinitions = definitions => definitions.some(definition => {
 		const scope = scopeManager.acquire(definition.node);
 		return isSameScope(parent, scope);

--- a/test/consistent-function-scoping.js
+++ b/test/consistent-function-scoping.js
@@ -256,7 +256,7 @@ ruleTester.run('consistent-function-scoping', rule, {
 
 				inner();
 			}
-		`,
+		`
 	],
 	invalid: [
 		{

--- a/test/consistent-function-scoping.js
+++ b/test/consistent-function-scoping.js
@@ -244,7 +244,19 @@ ruleTester.run('consistent-function-scoping', rule, {
 
 				eventEmitter.once('error', onError2);
 			};
-		`
+		`,
+		// #586
+		outdent`
+			function outer(stream) {
+				let content;
+
+				function inner() {
+					process.stdout.write(content);
+				}
+
+				inner();
+			}
+		`,
 	],
 	invalid: [
 		{

